Support matching a host in orbit enrollment using the serial number

mna opened the pull request:
**IMPORTANT:** Do no merge before 4.28.0 is released. 

#9124 

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [x] Changes file added for user-visible changes in `changes/` or `orbit/changes/`.
  See [Changes files](https://fleetdm.com/docs/contributing/committing-changes#changes-files) for more information.
- [x] Input data is properly validated, `SELECT *` is avoided, SQL injection is prevented (using placeholders for values in statements)
- [x] Added/updated tests

roperzh commented on orbit/cmd/orbit/orbit.go:
```diff
@@ -903,37 +905,44 @@ func (d *desktopRunner) interrupt(err error) {
+		HardwareSerial string `json:"hardware_serial"`
 	}
 
 	args := []string{"-S", "--json", "select uuid from system_info"}
```
```suggestion
	args := []string{"-S", "--json", "select uuid, hardware_serial from system_info"}
```

mna replied:
Oh man, that's embarrassing... :D Great catch!

roperzh reviewed (commented):
looks really good, left a requested change necessary for all to work together.

I did find it by DEP enrolling a computer, otherwise I would have approved lol. Maybe that log at the top is a bit confusing?

Once it's updated I will pull and test again

mna commented:
Failure is tuf-related, I'll merge once approved:

```
Error: initialize updates: failed to update metadata: update metadata: tuf: failed to download 3.root.json: Get "https://tuf.fleetctl.com/3.root.json": net/http: TLS handshake timeout
```

mna commented:
Roberto found another corner case that needs fixing, so we will merge this only in the next release. Details from the report, from slack:

what's happening so I don't forget: in practice the loop goes like this:

1. we get the device from ABM -> insert a host
2. EnrollOrbit is called
3. EnrollHost is called

3 is creating a duplicate row, I think we should be able to reproduce with an integration test.
I believe the reason is that `EnrollOrbit` sets the `osquery_host_id` when it does the `INSERT`, which is used by `EnrollOrbit` to try to find an existent host. Since the `INSERT` is not happening if the host comes from ABM, a duplicate row is created

mna commented:
Also to consider: it's probably not guaranteed that `EnrollOrbit` executes _before_ `EnrollOsquery` (unless orbit waits for the response of that call before starting the `osquery` process, I'll have to look into that).

mna commented:
@lukeheath @roperzh 

Ok so this was quite tricky to debug, but I think I have identified what is going on with `osquery_perf` and this change.

At the heart of it, it's an `osquery_perf` bug, because it always passes the same value as `hardware_serial` for all hosts, however it may have exposed an interesting case worth discussing.

So what happens is that when enrolling **without** orbit (and only then), host enrollment would result in failures in the calls to `/distributed/read` due to an `invalid node key`. The thing is that, because the serial is the same for all hosts, when a new host is enrolled (in EnrollAgent), it will look for a match by `osquery_host_id/uuid`, won't find any (as the uuid is unique per host and so it doesn't exist yet), but it _will_ find a match by serial number and will **update the existing host** to set its `osquery_node_id` and `node_key` to the values provided to the enrollment (even though it should've enrolled a new host).

On the other hand, when enrolling **with** orbit, it sets the `uuid` and `osquery_host_id` to the `hardware_uuid` received (which is unique per host as it should be in osquery_perf), so the subsequent osquery enrollment finds the proper node with the top-priority match by `osquery_host_id` (given that by default we use the `uuid` as osquery host identifier) and updates it with the osquery node key, causing no such `invalid node key` issue.

All that to say, I think the main fix is to provide unique serial numbers in `osquery_perf`, but I wonder if the second match by serial number could happen in real life (i.e. is it possible that non-unique serial numbers are used in the wild, maybe with VMs or such)? If so we'd run into similar issues there (wouldn't create new hosts if there's a pre-existing host with that serial).

(I'll double-check things to make sure this is what actually happens, but that's where my understanding of the thing is at the moment)

roperzh commented:
@mna thanks so much for that awesome explanation, a couple of thoughts:

> All that to say, I think the main fix is to provide unique serial numbers in osquery_perf

I sent a patch last week to use different serial numbers for osquery-perf: https://github.com/fleetdm/fleet/pull/9904, I'm just connecting the dots that this change was relevant to your work here, sorry about that! Do you know if you have that commit in your load test?

> I wonder if the second match by serial number could happen in real life (i.e. is it possible that non-unique serial numbers are used in the wild, maybe with VMs or such)? 

Yes, that's something we discussed in https://github.com/fleetdm/fleet/pull/8920#discussion_r1040941962. For now our assumption is that there might be duplicated serials if you use VMs, but that's alright at least for macOS.

mna commented:
@roperzh 

> Do you know if you have that commit in your load test?

I didn't as this is a relatively old branch and at first I wanted to keep it similar to the "before" test which was also an old-ish branch off of `main`, so implemented something similar but then saw that I had conflicts :D Anyway, now it has your changes and is up-do-date with the new commits as I didn't have the choice with the build/docker CI failures due to the node change.

> For now our assumption is that there might be duplicated serials if you use VMs, but that's alright at least for macOS

Ok, that complicates things a bit because we have to be able to match hosts by serial e.g. if it was created via DEP. Thinking out loud, we could do the lookup by serial only if:

* MDM is enabled and
* The matching host is macOS (well, darwin, as I think we save it)

This would also minimize the scope of the changes, for Fleet without MDM would still only match by `osquery_node_id`. Can you think of scenarios where that would be problematic?

roperzh commented:
@mna 

> so implemented something similar but then saw that I had conflicts :D

w00t! and sorry again! I wish I had mentioned that to you

> This would also minimize the scope of the changes, for Fleet without MDM would still only match by osquery_node_id. Can you think of scenarios where that would be problematic?

I can't think of anything and I like the idea 👍

mna commented:
@roperzh 

> and sorry again! I wish I had mentioned that to you

No worries, I'm pretty sure I saw it in passing when catching up on email and didn't connect the dots either as I didn't have that particular problem in mind.

mna commented:
@lukeheath @roperzh allright, I updated the PR as per the plan outlined in https://github.com/fleetdm/fleet/pull/9612#issuecomment-1440015309 and added tests to cover this new change. I will load-test this, and assuming it looks good, I'd suggest we QA and test this in dogfood during the next sprint (next week) and make it part of the next release (not this week's).

lukeheath commented:
@mna Thanks for the update and for digging into this. I agree; let's hold on merging this until after we release 4.28.0. Early next week, let's merge this in and deploy to dogfood for testing.

lukeheath commented:
I updated the PR description with that decision, so it doesn't accidentally get merged later this week.

mna commented:
@lukeheath @roperzh the results from the before/after with 10K hosts (after the duplicate serial number fix in `osquery-perf`. The graphs and behaviour are basically exactly the same (I'll omit the Redis graph as it's essentially identical and not very relevant). There's a slight difference in the DBs rows/s metric but that is because in the "after" scenario I started off with 8 load testers (4K hosts) before upping it to 10K (as I had to check something with the `osquery-perf` deployment), while in the "before" it was straight to 10K.

---
## Before

### Database Writer

![DB-writer-01](https://user-images.githubusercontent.com/692384/220763718-d9fbdbf0-0804-4b70-9cdc-a28ceb724a82.png)
![DB-writer-02](https://user-images.githubusercontent.com/692384/220763719-f894bd38-8623-46f6-8df9-a8f7c844c8eb.png)
![DB-writer-03](https://user-images.githubusercontent.com/692384/220763720-f7323d85-92b9-43af-a14c-5a63d70b6e3e.png)

### Database Replica

![DB-reader-01](https://user-images.githubusercontent.com/692384/220763891-d3063ba3-2934-4d2b-9dbd-dde9c5d97aff.png)
![DB-reader-02](https://user-images.githubusercontent.com/692384/220763894-9dca39b8-b8b3-4e08-a8a2-10376a41d5bc.png)
![DB-reader-03](https://user-images.githubusercontent.com/692384/220763895-e42f5e2c-9de7-49ce-be2e-655313c3dcd7.png)

### Fleet Instances

![ECS-Fleet-CPU-Memory](https://user-images.githubusercontent.com/692384/220763946-aec9f5f5-795a-4ba8-a0f9-51bb2dcd1faa.png)

### Random Notes

Before:

* with 20 loadtest instances
* 10K hosts show up in Fleet after a few minutes, all online, stable
* manage hosts page loads quickly, filtering by host status is fast
* live query (default one) proceeds on all hosts at ~100 results/s
* Fleet instances' logs show nothing about "cancelled", "context" or "locks"
* revision "2" of fleet and loadtest has been online since the start (i.e. after db migration executed), no unstable unhealthy instances
* loadtest instances see a bunch of "desktop received status 429 limit exceeded, retry after: 0s: limit exceeded, retry after: 0s" and "deviceClient.NumberOfFailingPolicies/CheckToken: unauthenticated, or invalid token", not sure if that's indication of any real problem
* loadtest error rate is very low/zero, except for desktop errors: "orbit errors: 237 desktop errors: 1552023-02-15 19:39:33.862940777 +0000 UTC m=+3900.094540796 :: error rate: 0.01 enrollments: 500 orbit enrollments: 262 writes: 866"

---
## After

### Database Writer

![DB-writer-01-02](https://user-images.githubusercontent.com/692384/220764295-0c74cfed-4b70-4506-8d49-0a7d062237e2.png)
![DB-writer-03](https://user-images.githubusercontent.com/692384/220764300-a688bdd1-304a-4a89-bade-24c787d2b78f.png)

### Database Replica

![DB-reader-01-02](https://user-images.githubusercontent.com/692384/220764358-bad5826d-94e8-4027-aad0-4fff86156061.png)
![DB-reader-03](https://user-images.githubusercontent.com/692384/220764362-3878e9aa-bdb7-40f9-8dc2-fe6cfc5514bf.png)

### Fleet Instances 

![ECS-Fleet-CPU-Memory](https://user-images.githubusercontent.com/692384/220764423-924b247e-68d7-4c5e-a9d6-72e896543d52.png)

### Random Notes

After (with duplicate serial numbers fixed in osquery-perf):

* with 20 loadtest instances
* 10K hosts show up in Fleet after a few minutes, all online, stable
* manage hosts page loads quickly, filtering by host status is fast
* live query (default one) proceeds on all hosts at ~100 results/s
* Fleet instances' logs show nothing about "cancelled", "context" or "locks"
* A few 'invalid node key" errors show up but it looks like it was at the start of the load test, after it settled at 10K hosts those errors did not show up anymore (there might've been such errors in the "before" test too though I didn't look for those as I did not know they needed to be searched for).
* revision "1" of fleet has been online since the start (i.e. after db migration executed), no unstable unhealthy instances
* loadtest instances see a bunch of "desktop received status 429 limit exceeded, retry after: 0s: limit exceeded, retry after: 0s" and "deviceClient.{NumberOfFailingPolicies,CheckToken}: unauthenticated, or invalid token", not sure if that's indication of any real problem
* loadtest error rate is very low/zero, except for desktop errors: "orbit errors: 237 desktop errors: 1552023-02-15 19:39:33.862940777 +0000 UTC m=+3900.094540796 :: error rate: 0.01 enrollments: 500 orbit enrollments: 262 writes: 866"

---

Results with 30K hosts coming in a moment (though there will not be a "before", but if everything still looks good at 30K I think there's not much concern for more testing).

roperzh reviewed (approved):
thanks so much for the in-depth testing and also for completing the serial-number stuff in osquery-perf 🎉

lukeheath commented:
Thanks for the in-depth update on this!
